**Carga y Limpieza de Datos - Lol e-sports**

**Importar las librerías**

In [7]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
pd.set_option('display.max_rows',200)
pd.set_option('display.max_columns',200)

**Importar el archivo CSV desde GitHub**

In [45]:
url = 'https://raw.githubusercontent.com/cporras/lol-esports/main/Data/2021_LoL_esports_match_data_from_OraclesElixir_20211020.csv'
df = pd.read_csv(url)

In [15]:
df.head()

gameid datacompleteness  \
0  ESPORTSTMNT03/1632489         complete   
1  ESPORTSTMNT03/1632489         complete   
2  ESPORTSTMNT03/1632489         complete   
3  ESPORTSTMNT03/1632489         complete   
4  ESPORTSTMNT03/1632489         complete   

                                                 url league  year split  \
0  http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2021   NaN   
1  http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2021   NaN   
2  http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2021   NaN   
3  http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2021   NaN   
4  http://matchhistory.na.leagueoflegends.com/en/...  KeSPA  2021   NaN   

   playoffs                 date  game  patch  playerid  side position  \
0         0  2021-01-02 07:40:39     1  10.25         1  Blue      top   
1         0  2021-01-02 07:40:39     1  10.25         2  Blue      jng   
2         0  2021-01-02 07:40:39     1  10.25         3  Blue      mid   
3         0  2021-01-02 07:40:39     1  10.25         4  Blue      bot   
4         0  2021-01-02 07:40:39     1  10.25         5  Blue      sup   

      player     team      champion    ban1    ban2      ban3      ban4  \
0       Khan  DWG KIA   Mordekaiser  Aatrox  Kai'Sa  Pantheon  Aphelios   
1     Canyon  DWG KIA        Graves  Aatrox  Kai'Sa  Pantheon  Aphelios   
2  ShowMaker  DWG KIA           Zoe  Aatrox  Kai'Sa  Pantheon  Aphelios   
3      Ghost  DWG KIA  Miss Fortune  Aatrox  Kai'Sa  Pantheon  Aphelios   
4      BeryL  DWG KIA         Galio  Aatrox  Kai'Sa  Pantheon  Aphelios   

     ban5  gamelength  result  kills  deaths  assists  teamkills  teamdeaths  \
0  Lucian        1782       1      7       1        3         23           4   
1  Lucian        1782       1      7       1        5         23           4   
2  Lucian        1782       1      5       0        8         23           4   
3  Lucian        1782       1      3       2       11         23           4   
4  Lucian        1782       1      1       0       14         23           4   

   doublekills  triplekills  quadrakills  pentakills  firstblood  \
0          0.0          0.0          0.0         0.0         1.0   
1          0.0          0.0          0.0         0.0         0.0   
2          1.0          0.0          0.0         0.0         0.0   
3          0.0          0.0          0.0         0.0         0.0   
4          0.0          0.0          0.0         0.0         0.0   

   firstbloodkill  firstbloodassist  firstbloodvictim  team kpm    ckpm  \
0             1.0               0.0               0.0    0.7744  0.9091   
1             0.0               0.0               0.0    0.7744  0.9091   
2             0.0               0.0               0.0    0.7744  0.9091   
3             0.0               0.0               0.0    0.7744  0.9091   
4             0.0               0.0               0.0    0.7744  0.9091   

   firstdragon  dragons  opp_dragons  elementaldrakes  opp_elementaldrakes  \
0          NaN      NaN          NaN              NaN                  NaN   
1          NaN      NaN          NaN              NaN                  NaN   
2          NaN      NaN          NaN              NaN                  NaN   
3          NaN      NaN          NaN              NaN                  NaN   
4          NaN      NaN          NaN              NaN                  NaN   

   infernals  mountains  clouds  oceans  dragons (type unknown)  elders  \
0        NaN        NaN     NaN     NaN                     NaN     NaN   
1        NaN        NaN     NaN     NaN                     NaN     NaN   
2        NaN        NaN     NaN     NaN                     NaN     NaN   
3        NaN        NaN     NaN     NaN                     NaN     NaN   
4        NaN        NaN     NaN     NaN                     NaN     NaN   

   opp_elders  firstherald  heralds  opp_heralds  firstbaron  barons  \
0         NaN          NaN      NaN          NaN         NaN     Na

In [16]:
df.shape

(138660, 119)

**Obtener la totalidad de registros por columnas**

In [17]:
df.count()

gameid                      138612
datacompleteness             96180
url                         138648
league                      138660
year                        138660
split                        92952
playoffs                    138660
date                        138660
game                        138660
patch                       138660
playerid                    138660
side                        138660
position                    138660
player                      115550
team                        138660
champion                    115550
ban1                        138270
ban2                        138390
ban3                        138108
ban4                        138264
ban5                        138000
gamelength                  138660
result                      138660
kills                       138660
deaths                      138660
assists                     138660
teamkills                   138660
teamdeaths                  138660
doublekills         

**Analisis del tipo de dato de cada columna**

In [18]:
df.dtypes

gameid                       object
datacompleteness             object
url                          object
league                       object
year                          int64
split                        object
playoffs                      int64
date                         object
game                          int64
patch                       float64
playerid                      int64
side                         object
position                     object
player                       object
team                         object
champion                     object
ban1                         object
ban2                         object
ban3                         object
ban4                         object
ban5                         object
gamelength                    int64
result                        int64
kills                         int64
deaths                        int64
assists                       int64
teamkills                     int64
teamdeaths                  

In [22]:
df['league'].value_counts().sort_values(ascending=False)

LDL        16344
LPL         8856
LCK         5868
LCSA        4980
RCL         4212
LCS         4140
NLC         4116
LMF         3840
OTBLX       3732
EM          3588
LHE         3456
PCS         3192
LEC         2964
LFL         2880
PRM         2856
SL          2724
LCK CL      2712
TCL         2688
CBLOL       2652
CBLOLA      2652
LVP DDH     2580
CU          2448
TRA         2448
UPL         2376
GLL         2376
PGN         2220
LLA         2208
LCO         2196
HM          2148
HC          2112
VCS         2112
UL          2028
LJL         1896
UKLC        1812
LCL         1788
BIG         1728
EBL         1692
LPLOL       1668
UGP         1584
BL          1236
DL          1212
NEXO        1188
WCS         1104
EGL         1104
BM          1056
MSI          972
HS           828
GSG          636
NERD         504
LJLA         468
AOL          444
KeSPA         36
Name: league, dtype: int64

**Se evalua usar solo las partidas de la ligas LPL o LCK, al ser estas ligas de alto nivel y con numerosos registros**. La liga LDL es la que tiene más registros, pero es una liga de segunda división

In [46]:
df_lck = df[df['league']=='LCK']
df_lpl = df[df['league']=='LPL']

In [48]:
df_lck.shape

(5868, 119)

In [49]:
df_lpl.shape

(8856, 119)

Los 8856 registros equivalen aproximadamente 738 partidas, debido a los 12 registros por partida que corresponden a los 5 jugadores de cada equipo mas 1 registro general por cada equipo

**Limpieza de datos**

In [53]:
#El dataframe tiene una columna que marca la completitud de los datos, nos quedamos solo con registros que tienen este flag
df_lck_complete = df_lck[df_lck['datacompleteness']=='complete']
df_lck_complete.shape

(3576, 119)

In [55]:
df_lpl_complete = df_lpl[df_lpl['datacompleteness']=='complete']
df_lpl_complete.shape

(4140, 119)

Se elige solo trabajar con datos marcados como completos de la liga LPL, corresponden a 345 partidas

In [103]:
#Existen columnas que son redundantes o no aportan a análisis que se quiere realizar, se eliminan del DF
df_cleaned = df_lpl_complete.drop(['datacompleteness','url','league','year','patch','ban1','ban2','ban3','ban4','ban5','elementaldrakes','opp_elementaldrakes','infernals','mountains','clouds','oceans','dragons (type unknown)','elders','opp_elders','damageshare','visionscore','vspm','earnedgoldshare','gspd'],axis=1)
df_cleaned.shape

(4140, 95)

In [81]:
df_cleaned.head(15)

gameid   split  playoffs                 date  game  playerid  side  \
36  6909-9183  Spring         0  2021-01-09 09:22:04     1         1  Blue   
37  6909-9183  Spring         0  2021-01-09 09:22:04     1         2  Blue   
38  6909-9183  Spring         0  2021-01-09 09:22:04     1         3  Blue   
39  6909-9183  Spring         0  2021-01-09 09:22:04     1         4  Blue   
40  6909-9183  Spring         0  2021-01-09 09:22:04     1         5  Blue   
41  6909-9183  Spring         0  2021-01-09 09:22:04     1         6   Red   
42  6909-9183  Spring         0  2021-01-09 09:22:04     1         7   Red   
43  6909-9183  Spring         0  2021-01-09 09:22:04     1         8   Red   
44  6909-9183  Spring         0  2021-01-09 09:22:04     1         9   Red   
45  6909-9183  Spring         0  2021-01-09 09:22:04     1        10   Red   
46  6909-9183  Spring         0  2021-01-09 09:22:04     1       100  Blue   
47  6909-9183  Spring         0  2021-01-09 09:22:04     1       200   Red   
48  6909-9184  Spring         0  2021-01-09 10:16:20     2         1  Blue   
49  6909-9184  Spring         0  2021-01-09 10:16:20     2         2  Blue   
50  6909-9184  Spring         0  2021-01-09 10:16:20     2         3  Blue   

   position      player         team  champion  gamelength  result  kills  \
36      top         369  Top Esports     Karma        2050       0      0   
37      jng       Karsa  Top Esports  Pantheon        2050       0      1   
38      mid      knight  Top Esports    Syndra        2050       0      2   
39      bot  JackeyLove  Top Esports    Samira        2050       0      0   
40      sup        Zhuo  Top Esports     Leona        2050       0      0   
41      top         Bin       Suning    Aatrox        2050       1      1   
42      jng        SofM       Suning   Nidalee        2050       1      1   
43      mid       Angel       Suning    Viktor        2050       1      2   
44      bot    huanfeng       Suning    Kai'Sa        2050       1      5   
45      sup          ON       Suning     Galio        2050       1      0   
46     team         NaN  Top Esports       NaN        2050       0      3   
47     team         NaN       Suning       NaN        2050       1      9   
48      top         369  Top Esports       Jax        2241       0      0   
49      jng       Karsa  Top Esports   Hecarim        2241       0      5   
50      mid      knight  Top Esports   Orianna        2241       0      8   

    deaths  assists  teamkills  teamdeaths  doublekills  triplekills  \
36       0        2          3           9          0.0          0.0   
37       2        1          3           9          0.0          0.0   
38       1        0          3           9          0.0          0.0   
39       2        2          3           9          0.0          0.0   
40       4        1          3           9          0.0          0.0   
41       0        6          9           3          0.0          0.0   
42       1        6          9           3          0.0          0.0   
43       0        5          9           3          0.0          0.0   
44       0        3          9           3          1.0          0.0   
45       2        8          9           3          0.0          0.0   
46       9        6          3           9          0.0          0.0   
47       3       28          9           3          1.0          0.0   
48       7        5         16          27          0.0          0.0   
49       6        6         16          27          1.0          0.0   
50       4        4         16          27          1.0          0.0   

    quadrakills  pentakills  firstblood  firstbloodkill  firstbloodassist  \
36          0.0         0.0         NaN             0.0               1.0   
37          0.0         0.0         NaN             0.0               0.0   
38          0.0         0.0         NaN             1.0               0.0   
39          0.0         0.0         NaN             0.0         

In [105]:
#Se concatenan las columnas gameid y game, este será el identificador de cada partida (Según la liga, un encuentro puede ser best of 3 o best of 5)
df_cleaned['gameid'] = df_cleaned['gameid'] + '-' + df_cleaned['game'].astype(str)
df_cleaned.shape

(4140, 95)

In [106]:
#Se desestiman los juegos sin ID
df_cleaned.dropna(subset=['gameid'],inplace=True)

#df_cleaned.head(15)
df_cleaned.shape

(4092, 95)

In [107]:
#Se renera el indice del DF
df_cleaned = df_cleaned.reset_index().drop('index',axis=1)

In [108]:
#Contaremos por el momento con dos DF, uno para las estadisticas de cada jugador y otro con el agregado por equipo
df_players = df_cleaned[df_cleaned['position']!='team']
df_players.head(10)

gameid   split  playoffs                 date  game  playerid  side  \
0  6909-9183-1-1  Spring         0  2021-01-09 09:22:04     1         1  Blue   
1  6909-9183-1-1  Spring         0  2021-01-09 09:22:04     1         2  Blue   
2  6909-9183-1-1  Spring         0  2021-01-09 09:22:04     1         3  Blue   
3  6909-9183-1-1  Spring         0  2021-01-09 09:22:04     1         4  Blue   
4  6909-9183-1-1  Spring         0  2021-01-09 09:22:04     1         5  Blue   
5  6909-9183-1-1  Spring         0  2021-01-09 09:22:04     1         6   Red   
6  6909-9183-1-1  Spring         0  2021-01-09 09:22:04     1         7   Red   
7  6909-9183-1-1  Spring         0  2021-01-09 09:22:04     1         8   Red   
8  6909-9183-1-1  Spring         0  2021-01-09 09:22:04     1         9   Red   
9  6909-9183-1-1  Spring         0  2021-01-09 09:22:04     1        10   Red   

  position      player         team  champion  gamelength  result  kills  \
0      top         369  Top Esports     Karma        2050       0      0   
1      jng       Karsa  Top Esports  Pantheon        2050       0      1   
2      mid      knight  Top Esports    Syndra        2050       0      2   
3      bot  JackeyLove  Top Esports    Samira        2050       0      0   
4      sup        Zhuo  Top Esports     Leona        2050       0      0   
5      top         Bin       Suning    Aatrox        2050       1      1   
6      jng        SofM       Suning   Nidalee        2050       1      1   
7      mid       Angel       Suning    Viktor        2050       1      2   
8      bot    huanfeng       Suning    Kai'Sa        2050       1      5   
9      sup          ON       Suning     Galio        2050       1      0   

   deaths  assists  teamkills  teamdeaths  doublekills  triplekills  \
0       0        2          3           9          0.0          0.0   
1       2        1          3           9          0.0          0.0   
2       1        0          3           9          0.0          0.0   
3       2        2          3           9          0.0          0.0   
4       4        1          3           9          0.0          0.0   
5       0        6          9           3          0.0          0.0   
6       1        6          9           3          0.0          0.0   
7       0        5          9           3          0.0          0.0   
8       0        3          9           3          1.0          0.0   
9       2        8          9           3          0.0          0.0   

   quadrakills  pentakills  firstblood  firstbloodkill  firstbloodassist  \
0          0.0         0.0         NaN             0.0               1.0   
1          0.0         0.0         NaN             0.0               0.0   
2          0.0         0.0         NaN             1.0               0.0   
3          0.0         0.0         NaN             0.0               0.0   
4          0.0         0.0         NaN             0.0               0.0   
5          0.0         0.0         NaN             0.0               0.0   
6          0.0         0.0         NaN             0.0               0.0   
7          0.0         0.0         NaN             0.0               0.0   
8          0.0         0.0         NaN             0.0               0.0   
9          0.0         0.0         NaN             0.0               0.0   

   firstbloodvictim  team kpm    ckpm  firstdragon  dragons  opp_dragons  \
0               0.0    0.0878  0.3512          NaN      NaN          NaN   
1               0.0    0.0878  0.3512          NaN      NaN          NaN   
2               0.0    0.0878  0.3512          NaN      NaN          NaN   
3               0.0    0.0878  0.3512          NaN      NaN          NaN   
4               0.0    0.0878  0.3512          NaN      NaN          NaN   
5               0.0    0.2634  0.3512          NaN      NaN          NaN   
6               1.0    0.2634  0.3512          NaN      NaN          NaN   
7               0.0    0.2634  0.3512          NaN    

In [109]:
#Contaremos por el momento con dos DF, uno para las estadisticas de cada jugador y otro con el agregado por equipo
df_teams = df_cleaned[df_cleaned['position']=='team']
df_teams.head(10)

gameid   split  playoffs                 date  game  playerid  \
10  6909-9183-1-1  Spring         0  2021-01-09 09:22:04     1       100   
11  6909-9183-1-1  Spring         0  2021-01-09 09:22:04     1       200   
22  6909-9184-2-2  Spring         0  2021-01-09 10:16:20     2       100   
23  6909-9184-2-2  Spring         0  2021-01-09 10:16:20     2       200   
34  6910-9189-1-1  Spring         0  2021-01-09 11:30:25     1       100   
35  6910-9189-1-1  Spring         0  2021-01-09 11:30:25     1       200   
46  6910-9190-2-2  Spring         0  2021-01-09 12:23:44     2       100   
47  6910-9190-2-2  Spring         0  2021-01-09 12:23:44     2       200   
58  6910-9191-3-3  Spring         0  2021-01-09 13:19:06     3       100   
59  6910-9191-3-3  Spring         0  2021-01-09 13:19:06     3       200   

    side position player           team champion  gamelength  result  kills  \
10  Blue     team    NaN    Top Esports      NaN        2050       0      3   
11   Red     team    NaN         Suning      NaN        2050       1      9   
22  Blue     team    NaN    Top Esports      NaN        2241       0     16   
23   Red     team    NaN         Suning      NaN        2241       1     27   
34  Blue     team    NaN      Oh My God      NaN        1887       0      4   
35   Red     team    NaN  EDward Gaming      NaN        1887       1     15   
46  Blue     team    NaN  EDward Gaming      NaN        2010       0      8   
47   Red     team    NaN      Oh My God      NaN        2010       1     18   
58  Blue     team    NaN      Oh My God      NaN        2159       0      8   
59   Red     team    NaN  EDward Gaming      NaN        2159       1     16   

    deaths  assists  teamkills  teamdeaths  doublekills  triplekills  \
10       9        6          3           9          0.0          0.0   
11       3       28          9           3          1.0          0.0   
22      27       27         16          27          2.0          0.0   
23      17       59         27          17          6.0          1.0   
34      15        9          4          15          0.0          0.0   
35       4       35         15           4          2.0          0.0   
46      18       18          8          18          0.0          0.0   
47       8       49         18           8          1.0          0.0   
58      16       19          8          16          1.0          0.0   
59       8       37         16           8          1.0          0.0   

    quadrakills  pentakills  firstblood  firstbloodkill  firstbloodassist  \
10          NaN         NaN         1.0             NaN               NaN   
11          NaN         NaN         0.0             NaN               NaN   
22          NaN         NaN         0.0             NaN               NaN   
23          NaN         NaN         1.0             NaN               NaN   
34          NaN         NaN         1.0             NaN               NaN   
35          NaN         NaN         0.0             NaN               NaN   
46          NaN         NaN         0.0             NaN               NaN   
47          NaN         NaN         1.0             NaN               NaN   
58          NaN         NaN         1.0             NaN               NaN   
59          NaN         NaN         0.0             NaN               NaN   

    firstbloodvictim  team kpm    ckpm  firstdragon  dragons  opp_dragons  \
10               NaN    0.0878  0.3512          1.0      2.0          2.0   
11               NaN    0.2634  0.3512          0.0      2.0          2.0   
22               NaN    0.4284  1.1513          0.0      3.0          2.0   
23               NaN    0.7229  1.1513          1.0      2.0          3.0   
34               NaN    0.1272  0.6041          1.0      1.0          4.0   
35               NaN    0.4769  0.6041          0.0      4.0          1.0   
46               NaN    0.2388  0.7761          0.0      1.0          4.0   
47               NaN    0.5373  0.7761         

Posteriormente se llevará el dataframe a un modelo de base de datos relacional y se agruparán los datos de jugadores y equipos

In [110]:
#Se exportan los dataframes para análisis siguientes
df_players.to_csv('df_players.csv',index=False)
df_teams.to_csv('df_teams.csv',index=False)